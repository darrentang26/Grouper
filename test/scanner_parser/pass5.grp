type z2 = Zero | One 

let z2 * z2 -> z2 add = match ((a,b)) with 
    ((Zero, Zero)) -> Zero
|   ((One, One))   -> Zero
|   _  -> One
and z2 -> z2 inverse = match (a) with 
    Zero -> One
|   One  -> Zero
and z2 identity = Zero
and z2 * z2 -> bool equals = function (n m) n == m  
and z2 group mod_2_group = group add identity inverse equals 
and z2 mult_id = One
and z2 * z2 -> z2 mult = match ((a,b)) with
    ((One, One))   -> One
|   _              -> Zero
and z2 ring mod_2_ring = ring mod_2_group mult mult_id