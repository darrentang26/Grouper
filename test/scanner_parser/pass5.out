type Z2 = ZERO | ONE
let (Z2 * Z2) -> Z2 add = (a, b) -> match (a b) with
  | (ZERO, ONE) -> ONE
  | (ONE, ZERO) -> ONE
  | (_) -> ZERO in
let Z2 -> Z2 inverse = (a) -> match (a) with
  | (ZERO) -> ONE
  | (ONE) -> ZERO in
let Z2 identity = ZERO in
let (Z2 * Z2) -> Bool equals = (n, m) -> n == m in
let Z2 group mod_2_group = Z2 add identity inverse equals in
let Z2 mult_id = ONE in
let (Z2 * Z2) -> Z2 mult = (a, b) -> match (a b) with
  | (ONE, ONE) -> ONE
  | (_) -> ZERO in
let Z2 ring mod_2_ring = Z2 add identity inverse equals mult mult_id in
ZERO + ONE == ZERO